cmd/cosign/cli/clean.go

```diff
@@ -62,32 +62,76 @@ func CleanCmd(ctx context.Context, regOpts options.RegistryOptions, cleanType op
 	}
 
 	remoteOpts := regOpts.GetRegistryClientOpts(ctx)
+	ociRemoteOpts := ociremote.WithRemoteOptions(remoteOpts...)
 
-	sigRef, err := ociremote.SignatureTag(ref, ociremote.WithRemoteOptions(remoteOpts...))
+	sigRef, err := ociremote.SignatureTag(ref, ociRemoteOpts)
 	if err != nil {
 		return err
 	}
 
-	attRef, err := ociremote.AttestationTag(ref, ociremote.WithRemoteOptions(remoteOpts...))
+	attRef, err := ociremote.AttestationTag(ref, ociRemoteOpts)
 	if err != nil {
 		return err
 	}
 
-	sbomRef, err := ociremote.SBOMTag(ref, ociremote.WithRemoteOptions(remoteOpts...))
+	sbomRef, err := ociremote.SBOMTag(ref, ociRemoteOpts)
 	if err != nil {
 		return err
 	}
 
-	var cleanTags []name.Tag
+	referrerRefs := []name.Reference{}
+	digest, ok := ref.(name.Digest)
+	if !ok {
+		var err error
+		digest, err = ociremote.ResolveDigest(ref, ociRemoteOpts)
+		if err != nil {
+			return fmt.Errorf("resolving digest: %w", err)
+		}
+	}
+	idx, err := remote.Referrers(digest, remoteOpts...)
+	if err != nil {
+		return err
+	}
+	if idx != nil {
+		// Delete manifest
+		imgDigest, err := idx.Digest()
+		if err != nil {
+			return err
+		}
+		referrerDigestStr := fmt.Sprintf("%s@%s", ref.Context().Name(), imgDigest.String())
+		referrerDigest, err := name.NewDigest(referrerDigestStr)
+		if err != nil {
+			return err
+		}
+		referrerRefs = append(referrerRefs, referrerDigest)
+
+		// Delete layers in the manifest
+		idxManifest, err := idx.IndexManifest()
+		if err != nil {
+			return err
+		}
+		if idxManifest != nil {
+			for _, manifest := range idxManifest.Manifests {
+				layerDigestStr := fmt.Sprintf("%s@%s", ref.Context().Name(), manifest.Digest.String())
+				layerDigest, err := name.NewDigest(layerDigestStr)
+				if err != nil {
+					return err
+				}
+				referrerRefs = append(referrerRefs, layerDigest)
+			}
+		}
+	}
+
+	var cleanTags []name.Reference
 	switch cleanType {
```

Comment:
It seems like `--type` no longer has any meaning when it comes to referrers - when I sign with `sign` but then clean with `--type=attestation`, it just removes everything. That could come as a surprise to someone who's expecting those objects to be different.

Reply:
Yeah, unfortunately with the new protobuf bundle (where we're always trying to use the OCI1.1 referrer specification) both [attest](https://github.com/sigstore/cosign/blob/4e3042cebab7fefd0a64f803979b6cfb00859ddd/cmd/cosign/cli/attest/attest.go#L127) and [sign](https://github.com/sigstore/cosign/blob/4e3042cebab7fefd0a64f803979b6cfb00859ddd/cmd/cosign/cli/sign/sign.go#L192) are using `CosignSignPredicateType`, so I don't think there's an easy way to tell them apart.

I think eventually we might consider merging `sign` and `attest`, but of course that isn't where we are today.

Alternatively, we could only clean up referring items on `CleanTypeAll`? That could also be confusing. 🤷 Definitely open to feedback here.

Reply:
What if there was a new `--type=referrer` just for this? Then we could print a warning if referrers are found but a signature or attestation type is specified to let the user know they need to use specify the new type. I think at some point we just need to accept that referrers are different and can't be cajoled into behaving like classic `.sig`s or `.att`s.